File: backend/setu/bucket_lineage_adapter.py
from datetime import datetime
from typing import Any, Dict, Optional

from .mongo_store import MongoSetuStore
from .utils import compute_determinism_hash


class BucketLineageAdapter:
    def __init__(self, store: MongoSetuStore):
        self.store = store

    async def emit_execution_event(self, execution: Dict[str, Any], event_type: str,
                                   payload: Optional[Dict[str, Any]] = None,
                                   overrides: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        if not execution.get("execution_id") or not execution.get("trace_id") or not execution.get("tenant_id"):
            raise ValueError("Execution identifiers are required for lineage emission")

        overrides = overrides or {}
        timestamp = overrides.get("timestamp") or execution.get("timestamp") or datetime.utcnow().isoformat()
        sequence = overrides.get("sequence") or await self.store.next_lineage_sequence(execution.get("trace_id"))

        event = {
            "lineage_event_id": "",
            "execution_id": execution.get("execution_id"),
            "trace_id": execution.get("trace_id"),
            "tenant_id": execution.get("tenant_id"),
            "event_type": event_type,
            "timestamp": timestamp,
            "sequence": sequence,
            "payload": payload or {}
        }

        event["determinism_hash"] = compute_determinism_hash(event)
        event["lineage_event_id"] = "lin_" + event["determinism_hash"][:16]

        await self.store.append_lineage_event(event)
        return event

    async def list_events(self, trace_id: str, limit: int = 200):
        return await self.store.list_lineage_events(trace_id, limit)

File: backend-nodejs/src/services/bucketLineageAdapter.js
import crypto from 'crypto';
import mongoose from 'mongoose';

/**
 * bucketLineageAdapter.js — SETU Bucket Lineage Adapter (Node.js Edition)
 *
 * Emits and verifies deterministic lineage events across execution traces.
 */

// MongoDB Schema for SETU Lineage Events
const setuLineageEventSchema = new mongoose.Schema({
  lineage_event_id: { type: String, required: true, unique: true },
  execution_id: { type: String, required: true, index: true },
  trace_id: { type: String, required: true, index: true },
  tenant_id: { type: String, required: true, index: true },
  event_type: { type: String, required: true },
  timestamp: { type: String, required: true },
  sequence: { type: Number, required: true },
  payload: { type: Object, default: {} },
  determinism_hash: { type: String, required: true },
}, { timestamps: true });

const SetuLineageEvent = mongoose.models.SetuLineageEvent || mongoose.model('SetuLineageEvent', setuLineageEventSchema);

/**
 * Computes deterministic SHA-256 hash for lineage event payload
 */
export function computeDeterminismHash(eventObj) {
  const content = `${eventObj.execution_id}|${eventObj.trace_id}|${eventObj.tenant_id}|${eventObj.event_type}|${eventObj.timestamp}|${eventObj.sequence}`;
  return crypto.createHash('sha256').update(content).digest('hex');
}

export class BucketLineageAdapter {
  /**
   * Emits a lineage event with deterministic hash and sequence counter
   */
  static async emitExecutionEvent(execution, eventType, payload = {}, overrides = {}) {
    if (!execution || !execution.execution_id || !execution.trace_id || !execution.tenant_id) {
      throw new Error('Execution identifiers (execution_id, trace_id, tenant_id) are required for lineage emission');
    }

    const timestamp = overrides.timestamp || execution.timestamp || new Date().toISOString();
    
    let sequence = overrides.sequence || 1;
    if (mongoose.connection.readyState === 1) {
      const lastEvent = await SetuLineageEvent.findOne({ trace_id: execution.trace_id }).sort({ sequence: -1 }).lean();
      if (lastEvent) {
        sequence = lastEvent.sequence + 1;
      }
    }

    const eventObj = {
      execution_id: execution.execution_id,
      trace_id: execution.trace_id,
      tenant_id: execution.tenant_id,
      event_type: eventType,
      timestamp: timestamp,
      sequence: sequence,
      payload: payload || {},
    };

    const determinismHash = computeDeterminismHash(eventObj);
    const lineageEventId = `lin_${determinismHash.slice(0, 16)}`;

    const finalEvent = {
      lineage_event_id: lineageEventId,
      ...eventObj,
      determinism_hash: determinismHash,
    };

    if (mongoose.connection.readyState === 1) {
      await SetuLineageEvent.create(finalEvent);
    }

    return finalEvent;
  }

  /**
   * List lineage events by trace_id
   */
  static async listEvents(traceId, limit = 200) {
    if (mongoose.connection.readyState === 1) {
      return await SetuLineageEvent.find({ trace_id: traceId }).sort({ sequence: 1 }).limit(limit).lean();
    }
    return [];
  }

  /**
   * Verify execution event history exists in lineage store
   */
  static async verifyExecutionHistory(executionId, traceId) {
    const events = await this.listEvents(traceId, 1000);
    const eventFound = events.some(e => e.execution_id === executionId);

    return {
      execution_id: executionId,
      trace_id: traceId,
      verified: eventFound,
      verification_details: {
        execution_event_exists: eventFound,
        total_trace_events: events.length,
      },
    };
  }
}

export default BucketLineageAdapter;
